src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixMult.java

```diff
@@ -123,6 +126,52 @@ public static void matrixMult(MatrixBlock m1, MatrixBlock m2, MatrixBlock ret, i
 	{
 		matrixMult(m1, m2, ret, rl, ru, true);
 	}
+	
+	// We could encapsulate heuristics in this function
+	// For now, we only consider matrix-vector operation to be memory bound
+	private static boolean isMatMultMemoryBound(int m1Rlen, int m1Clen, int m2Clen) {
+		if(m1Rlen == 1 || m1Clen == 1 || m2Clen == 1)
+			return true;
+		else
+			return false;
+	}
+	
+	/**
+	 * Performs matrix multiplication using native library if BLAS is available or else falls back to Java BLAS.
+	 * 
+	 * @param m1 lhs matrix block
+	 * @param m2 rhs matrix block
+	 * @param ret output matrix block
+	 * @param k number of threads
+	 * @throws DMLRuntimeException if error occurs
+	 */
+	public static void matrixMultNative(MatrixBlock m1, MatrixBlock m2, MatrixBlock ret, int k) throws DMLRuntimeException {	
+			//check inputs / outputs
+			if( m1.isEmptyBlock(false) || m2.isEmptyBlock(false) ) {
+				ret.examSparsity(); //turn empty dense into sparse
+				return;
+			}
+			if(DMLScript.isNativeEnabled(k) && !isMatMultMemoryBound(m1.rlen, m1.clen, m2.clen) && !m1.isInSparseFormat() && !m2.isInSparseFormat()) {
+				ret.sparse = false;
+				ret.allocateDenseBlock();
+				if(NativeHelper.matrixMultDenseDense(m1.denseBlock, m2.denseBlock, ret.denseBlock, m1.getNumRows(), 
+						m1.getNumColumns(), m2.getNumColumns(), k > 0 ? k : NativeHelper.getMaxNumThreads())) {
```

Comment:
please remove the entire integration from LibMatrixMult in order to simplify isolating our matrix block library as a pure java library.